src/microengine_utils/scanner.py

```diff
@@ -0,0 +1,165 @@
+import asyncio
+import functools
+import platform
+import time
+from contextlib import contextmanager, suppress
+from typing import (TYPE_CHECKING, Callable, NewType, Optional, Text, Tuple, TypedDict, Union)
+
+import datadog
+from polyswarmartifact import ArtifactType
+from polyswarmartifact.schema.verdict import Verdict
+
+from .datadog import (SCAN_FAIL, SCAN_NO_RESULT, SCAN_SUCCESS, SCAN_TIME, SCAN_TYPE_INVALID, SCAN_VERDICT)
+from .errors import (BaseScanError, CalledProcessScanError, MalformedResponseScanError, TimeoutScanError)
+from .fs import AsyncArtifactTempfile, winepath
+
+if TYPE_CHECKING:
+    ExitCodeT = int
+
+    from polyswarmclient import ScanResult
+    from datetime import datetime
+
+    FILE_TYPE = ArtifactType.FILE
+    URL_TYPE = ArtifactType.URL
+
+
+async def create_scanner_exec(
+    *cmd: 'str',
+    stdout: 'asyncio.StreamReader' = asyncio.subprocess.PIPE,
+    stderr: 'asyncio.StreamReader' = asyncio.subprocess.PIPE,
+    stdin: 'asyncio.StreamReader' = asyncio.subprocess.DEVNULL,
+    check: 'bool' = False,
+    timeout: 'int' = 15,
+) -> 'Tuple[ExitCodeT, Text, ...]':
+    """Run an engine filescan `cmd`, timing out after `timeout` seconds"""
+    try:
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            *cmd,
+            stdout=stdout,
+            stderr=stderr,
+            stdin=stdin,
+        )
+        streams = await asyncio.wait_for(proc.communicate(), timeout=timeout)
+        if check and proc.returncode != 0:
+            raise CalledProcessScanError
+        return (
+            proc.returncode,
+            *(s.decode(errors='ignore') if isinstance(s, bytes) else str(s or '') for s in streams)
+        )
+    except (BrokenPipeError, ConnectionResetError):  # noqa
+        proc.kill()
+        raise CalledProcessScanError
+    except asyncio.TimeoutError as e:
+        proc.kill()
+        raise TimeoutScanError
+
+
+def collect_scan(verbose: 'bool' = False) -> 'Callable[..., ScanResult]':
+    """Decorator for `async_scan` to automatically handle errors and boilerplate scanner metadata
```

Comment:
if this is for `async_scan` should we call it `collect_async_scan`? And should we have a similar function for the `sync_scan` ?